Added Master/Available cluster metrics

maxsumrall opened the pull request:
Implemented metrics reporting the IS_MASTER and IS_AVAILABLE status of a cluster member.

Commit:
Added Master/Available cluster metrics

Implemented metrics reporting the IS_MASTER and IS_AVAILABLE status of a cluster member.

chrisvest commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsSettings.java:
```diff
@@ -81,6 +81,10 @@
     public static Setting<Boolean> jvmThreadsEnabled = setting( "metrics.jvm.threads.enabled", Settings.BOOLEAN, neoEnabled );
 
+    public static Setting<? extends Boolean> clusterEnabled = setting( "metrics.neo4j.cluster.enabled", Settings
+            .BOOLEAN, neoEnabled);
```
Line-break after the comma. It looks better.

chrisvest commented on enterprise/metrics/src/test/java/org/neo4j/metrics/source/ClusterMetricsTest.java:
```diff
@@ -0,0 +1,128 @@
+        reporter.start( 10, TimeUnit.MILLISECONDS );
+
+        // then wait for the reporter to get a report
+        reporter.awaitReport();
```
Call the `report()` method of the `ScheduledReporter` instead of starting a background thread (that you don't stop) and waiting for it to do stuff.

chrisvest commented on enterprise/metrics/src/test/java/org/neo4j/metrics/source/ClusterMetricsTest.java:
```diff
@@ -0,0 +1,128 @@
+
+    }
+
+}
```
Some line-break cleanup would be good here.

chrisvest commented:
@maxsumrall Had a few comments but it mostly looks good. This will conflict with #5773

maxsumrall commented:
Waiting for #5778 and #5773

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/MetricsSettings.java:
```diff
@@ -81,6 +81,10 @@
     @Description( "Enable reporting metrics about the current number of threads running on the HA cluster component." )
     public static Setting<Boolean> jvmThreadsEnabled = setting( "metrics.jvm.threads.enabled", Settings.BOOLEAN, neoEnabled );
 
+    public static Setting<? extends Boolean> clusterEnabled = setting( "metrics.neo4j.cluster.enabled", Settings
```
We need to have description for this setting.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -0,0 +1,92 @@
+public class ClusterMetrics extends LifecycleAdapter
+{
+    static final String NAME_PREFIX = "neo4j.cluster";
+    static final String IS_MASTER = name( NAME_PREFIX, "is_master" );
```
we need to describe exposed metrics, so they can be included into documentation, so people can understand when they can use them

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -0,0 +1,92 @@
+    {
+        private Predicate<String> rolePredicate;
+
+        public RoleGauge( Predicate<String> rolePredicate )
```
this can be simplified just to use plain role string instead of wrapping into predicate

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -0,0 +1,92 @@
+            this.rolePredicate = rolePredicate;
+        }
+
+        public Integer getValue()
```
why we converting result to integer here instead of just returning boolean?
Is it interface agreement that it should be always number type?

chrisvest replied:
It makes it easier to work with in charts, I guess.

MishaDemianenko replied:
That is my assumption as well, but if charts can handle that that would be also nice, just to know that we know what we are doing

MishaDemianenko commented on enterprise/metrics/src/test/java/org/neo4j/metrics/source/ClusterMetricsTest.java:
```diff
@@ -0,0 +1,128 @@
+import static org.neo4j.helpers.collection.MapUtil.stringMap;
+
+
+public class ClusterMetricsTest
```
Test always checking happy path when `MetricsSettings.clusterEnabled` is true, should have simple test when it's false

MishaDemianenko commented:
@maxsumrall  https://github.com/neo4j/neo4j/pull/5778 was merged, can you please also rebase

Commit:
Added Master/Available cluster metrics

Changed cluster metrics to use a dependency resolver to allow for community builds to build the metrics package without causing problems since these metrics actually need enterprise/HA features.

maxsumrall commented:
@MishaDemianenko 
Rebased and fixed conflicts. Please review again as much has changed.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -75,6 +96,30 @@ public Long getValue()
+
+    private void registerClusterMemberMetrics()
+    {
+        if(clusterMembers != null)
```
code style

MishaDemianenko replied:
we already checking availability of clusterMembers before invoking this method and will not invoke registration if we failed at all

maxsumrall replied:
Yes, this null check is not needed.

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -56,6 +72,11 @@ public void start() throws Throwable
     {
         if ( config.get( MetricsSettings.neoClusterEnabled ) )
         {
+            if(canResolveClusterMembersDependency())
```
code style

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -75,6 +96,30 @@ public Long getValue()
+        {
+            clusterMembers = dependencyResolver.resolveDependency( ClusterMembers.class );
+        }
+        catch(IllegalArgumentException e)
```
here as well

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -85,6 +130,8 @@ public void stop() throws IOException
         {
             registry.remove( SLAVE_PULL_UPDATES );
             registry.remove( SLAVE_PULL_UPDATE_UP_TO_TX );
+            registry.remove( IS_MASTER );
```
should we check if we registered them?

MishaDemianenko commented on enterprise/metrics/src/test/java/org/neo4j/metrics/MetricsKernelExtensionFactoryIT.java:
```diff
@@ -88,6 +89,7 @@ public void mustLoadMetricsExtensionWhenConfigured() throws Exception
                 tx.success();
             }
         }
+        cluster.stop();
```
why do we stop cluster here?

maxsumrall replied:
At this point the cluster is not needed. The rest of that test is only looking at the output log file. It's good to cleanup your mess closer to the source.

MishaDemianenko replied:
was wondering just because it does not much other tests :)

MishaDemianenko commented on enterprise/metrics/src/test/java/org/neo4j/metrics/source/ClusterMetricsTest.java:
```diff
@@ -0,0 +1,181 @@
+        reporter.start( 10, TimeUnit.MILLISECONDS );
+
+        //then the reporter should fail
+        try
```
expectedException rule?

MishaDemianenko commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -75,6 +96,30 @@ public Long getValue()
+        }
+    }
+
+    private boolean canResolveClusterMembersDependency()
```
it would be really nice to have test that verify behaviour in case when there is no cluster members available

maxsumrall replied:
@MishaDemianenko  Yes, I added that.

Commit:
Cluster Metrics for cluster member role.

Added additional test for when cluster metrics are not available but reporters attempt to get those values. Previously a NullPointerException was thrown. Now those metrics just return a negative response.

MishaDemianenko commented:
@maxsumrall lets cleanup tomorrow test code style a bit otherwise fine, will merge it tomorrow

Commit:
Additional improvements to cluster metrics for IS_MASTER and IS_AVAILABLE

davidegrohmann commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -75,9 +92,30 @@ public Long getValue()
+        }
+    }
+
+    private void resolveClusterMembersDependencyOrLogWarning()
```
this should return a boolean telling us if the resolution was successful or not

davidegrohmann commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -56,6 +72,7 @@ public void start() throws Throwable
     {
         if ( config.get( MetricsSettings.neoClusterEnabled ) )
```
this should be `if ( config.get( MetricsSettings.neoClusterEnabled && resolveClusterMembersDependencyOrLogWarning )`

davidegrohmann replied:
this will actually make the old code better, since we'll register the `SLAVE_PULL_UPDATES` metrics only if we are in cluster mode.

davidegrohmann commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -56,6 +72,7 @@ public void start() throws Throwable
     {
         if ( config.get( MetricsSettings.neoClusterEnabled ) )
         {
+            registerClusterMemberMetrics();
```
this can be inlined by removing the call to `resolveClusterMembersDependencyOrLogWarning`

davidegrohmann commented on enterprise/metrics/src/main/java/org/neo4j/metrics/source/ClusterMetrics.java:
```diff
@@ -86,6 +124,12 @@ public void stop() throws IOException
             registry.remove( SLAVE_PULL_UPDATES );
             registry.remove( SLAVE_PULL_UPDATE_UP_TO_TX );
 
+            if( clusterMembers != null )
```
this condition should be moved up to the other if, i.e., `if ( config.get( MetricsSettings.neoClusterEnabled ) && clusterMembers != null )`

maxsumrall commented:
We have decided to make these changes against 3.0.
